src/Solve.cpp

```
@@ -0,0 +1,132 @@
+  switch (solver_options.id) {
+    case BLOCK_DIAGONAL: {
+      std::shared_ptr<PreconditionBlockDiagonal> actual_precondition =
+          std::make_shared<PreconditionBlockDiagonal>();
+      actual_precondition->initialize(
+          system_matrix.block(0, 0), pressure_mass.block(1, 1),
+          solver_options.maxiter_inner, solver_options.tol_inner,
+          solver_options.use_ilu);
+      precondition = actual_precondition;
+      break;
+    }
+    case SIMPLE: {
+      std::shared_ptr<PreconditionSIMPLE> actual_precondition =
+          std::make_shared<PreconditionSIMPLE>();
+      actual_precondition->initialize(
+          system_matrix.block(0, 0), system_matrix.block(1, 0),
+          system_matrix.block(0, 1), solution_owned, solver_options.alpha,
+          solver_options.maxiter_inner, solver_options.tol_inner,
+          solver_options.use_ilu);
+      precondition = actual_precondition;
+      break;
+    }
+    case ASIMPLE: {
+      std::shared_ptr<PreconditionaSIMPLE> actual_precondition =
+          std::make_shared<PreconditionaSIMPLE>();
+      actual_precondition->initialize(
+          system_matrix.block(0, 0), system_matrix.block(1, 0),
+          system_matrix.block(0, 1), solution_owned, solver_options.alpha,
+          solver_options.use_inner_solver, solver_options.maxiter_inner,
+          solver_options.tol_inner, solver_options.use_ilu);
+      precondition = actual_precondition;
+      break;
+    }
+    case YOSHIDA: {
+      std::shared_ptr<PreconditionYoshida> actual_precondition =
+          std::make_shared<PreconditionYoshida>();
+      actual_precondition->initialize(
+          system_matrix.block(0, 0), system_matrix.block(1, 0),
+          system_matrix.block(0, 1), velocity_mass.block(0, 0), solution_owned,
+          solver_options.maxiter_inner, solver_options.tol_inner,
+          solver_options.use_ilu);
+      precondition = actual_precondition;
+      break;
+    }
+  }
```

If I understand correctly, you use two pointers (one to the base class, one to the derived class) because the base class does not have an `initialize` method.

However, seeing as you call `initialize` right after constructing the object, you could just merge the `initialize` method with the constructor of each derived class (the constructors of classes derived from the same base can have different signatures). This way, for every `case` statement you would only have to write
```cpp
case BLOCK_DIAGONAL:
  preconditioner = std::make_shared<PreconditionBlockDiagonal>( /* ...args... */);
  break;
```

A couple of side notes:
- you should not use a `shared_ptr` when a `unique_ptr` would do (as is the case here);
- your `switch` statements should always have a `default` case, and if you have no way of handling the default case you should throw an exception.